packages/fxa-auth-server/scripts/refund-unverified-accounts.ts

```diff
@@ -0,0 +1,399 @@
+}
+
+/**
+ * Refund and Delete Customer, Cancel Subscription
+ */
+export async function cancelSubscriptionsAndDeleteCustomer(
+  uid: string,
+  email: string,
+  subscriptionId: string,
+  invoiceId: string
+): Promise<void> {
+  const stripeHelper = Container.get(StripeHelper);
+  const firestore = Container.get(AuthFirestore);
+
+  const invoice = await retrieveInvoice(firestore, invoiceId);
+  const paymentIntent =
+    typeof invoice.payment_intent === 'string'
+      ? invoice.payment_intent
+      : invoice.payment_intent?.id;
+
+  if (!paymentIntent) {
+    throw new Error(`Could not find payment intent for ${email}`);
+  }
+
+  // Cancel any subscriptions and issue refunds
+  await stripeHelper.refundPayment(paymentIntent, 'fraudulent');
+  await stripeHelper.cancelSubscription(subscriptionId);
+
+  // Remove the customer from stripe
+  await stripeHelper.removeCustomer(uid, email);
+}
+
+async function mergeAccountData(database: any, subscription: SubscriptionData) {
+  const { subscription: sub, uid } = subscription;
+  try {
+    const account = await database.account(uid);
+    if (account) {
+      const outputData: OutputData = {
+        uid,
+        email: account.email,
+        accountCreated: account.createdAt,
+        customerId: sub.customer as string,
+        subscriptionId: sub.id,
+        subscriptionCreated: sub.created,
+        subscriptionStatus: sub.status,
+        subscriptionPlan: sub.items.data[0].id,
+        subscriptionInvoiceId: sub.latest_invoice as string,
+        refunded: false,
+      };
+      if (account.verifierSetAt === 0) {
+        return { unverifiedAccount: outputData };
+      } else {
+        return { verifiedAccount: outputData };
+      }
+    } else {
+      throw new Error('Could not find FxA account');
+    }
+  } catch (error) {
+    throw new MergedAccountError(error.message, subscription.uid);
+  }
+}
+
+async function refundUnverifiedAccounts({
+  database,
+  isDryRun,
+  startDate,
+  endDate,
+  skipFile,
+}: {
+  database: any;
+  isDryRun: boolean;
+  startDate: number;
+  endDate: number;
+  skipFile: string;
+}) {
+  const unverifiedAccounts: OutputData[] = [];
+  const unverifiedAccountsRefunded: OutputData[] = [];
+  const verifiedAccounts: OutputData[] = [];
+  const errorsOutput: ErrorOutput[] = [];
+
+  const skipUidOrEmail = skipFile ? await readSkip(skipFile) : [];
+
+  try {
+    // Retrieve all subscriptions, and related account UID from Firestore, for specified date range.
+    const subscriptions: SubscriptionData[] =
+      await retrieveSubscriptionsByCreated(startDate, endDate);
+
+    // For each UID, get the Account record
+    // Populate either unverified or verified Account arrays
+    // Or errorsOutput if an error ocurred.
+    const mergedDataResults = await Promise.allSettled(
+      subscriptions.map((sub) => mergeAccountData(database, sub))
+    );
+    mergedDataResults.forEach((acc) => {
+      if (acc.status === 'fulfilled') {
```

Comment:
I am not sure what `acc.status` is. It doesn't look like its part of the `OutputData` that is returned from `mergeAccountData` and `fulfilled` isn't a stripe subscription status. I tried searching for `fulfilled` in our code briefly but didn't find anything illuminating.

Not saying this is wrong, but was wondering if you might be able to elaborate herefor my own benefit? I am assuming you have tested this locally and it works hence why I am asking.

Reply:
In this case, I was using `Promise.allSettled` which returns a status either as `fulfilled` or `rejected` for each promise, instead of using `Promise.all` which will reject if any of the promises reject. (Better docs and explanation can be found here [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled))